### PR TITLE
Better member annotation scoping

### PR DESCRIPTION
**Pull request by MDoerner**
This PR simplifies the annotation scoping.

After this PR, the rules will be as follows, where annotations and comments are considered to be whitespace.

- Annotations on non-whitespace logical lines are illegal and do not annotate anything.
- Annotations annotate the next physical non-whitespace line below them.
- Whether an annotation annotates a construct on the annotated line depends on its `AnnotationType`.
- Module annotations annotate the module, provided the annotated line is at most the first line of the first module body element, if there is any.

In addition, this PR changes the `GetDescendent<TContext>()` extension to a depth first search approach from traversing the entire tree collecting all candidates, then sorting them and finally taking the first.

### CODE REVIEW

File: Rubberduck.CodeAnalysis/Inspections/Abstract/InspectionBase.cs [hunk 1]
```diff
@@ -86,28 +86,22 @@ protected virtual IEnumerable<Declaration> BuiltInDeclarations
...
-                var ignoreModuleAnnotation = annotation as IgnoreModuleAnnotation;
-
-                if (ignoreAnnotation?.InspectionNames.Contains(AnnotationName) == true)
+                if (ignoreAnnotation.InspectionNames.Contains(AnnotationName))
                 {
                     return true;
                 }
+            }
+
+            var moduleDeclaration = State.DeclarationFinder.Members(module)
```
Comment (retailcoder) [inline Rubberduck.CodeAnalysis/Inspections/Abstract/InspectionBase.cs]: Wouldn't it be more efficient to check for module-level annotations first? Worst-case would be a massive module with many members, right?

File: Rubberduck.Parsing/Annotations/IdentifierAnnotationService.cs [hunk 2]
```diff
@@ -16,24 +16,7 @@ public IdentifierAnnotationService(DeclarationFinder declarationFinder)
...
-                {
-                    break;
-                }
-
-                var annotationsStartingOnCurrentLine = moduleAnnotations.Where(a => a.QualifiedSelection.Selection.StartLine == currentLine && a.AnnotationType.HasFlag(AnnotationType.IdentifierAnnotation));
-
-                annotations.AddRange(annotationsStartingOnCurrentLine);
-            }
-            return annotations;
+            return _declarationFinder.FindAnnotations(module, line).Where(annotation => annotation.AnnotationType.HasFlag(AnnotationType.IdentifierAnnotation));
```
Comment (retailcoder) [inline Rubberduck.Parsing/Annotations/IdentifierAnnotationService.cs]: Nice! Should this method be inlined at this point?

File: Rubberduck.CodeAnalysis/Inspections/Abstract/InspectionBase.cs (hunk 1 above)
Comment (MDoerner) [reply Rubberduck.CodeAnalysis/Inspections/Abstract/InspectionBase.cs]: Sorry, I do not understand what you mean. The part above looking for `IgnoreAnnotations` for the line is a dictionary lookup. Where do the members enter, here?

File: Rubberduck.Parsing/Annotations/IdentifierAnnotationService.cs (hunk 2 above)
Comment (MDoerner) [reply Rubberduck.Parsing/Annotations/IdentifierAnnotationService.cs]: That would make sense. Then, the entire class is needed no longer.

**Pull request merged**